### PR TITLE
Add data export endpoint

### PR DESCRIPTION
**Pull request by roborourke**
Add the endpoint `/wp-json/analytics/v1/events/YYYY-MM-DD` - this allows exporting data for the given day.

Pagination info is delivered via response headers.

Still need to add docs around using app passwords for auth.

### CODE REVIEW

File: inc/export/class-endpoint.php [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+						'description' => 'Day to retrieve the analytics data from.',
+						'type' => 'date',
+						'sanitize_callback' => 'sanitize_text_field',
+					],
+					'page' => [
+						'description'       => 'Current page',
+						'type'              => 'integer',
+						'default'           => 1,
+						'sanitize_callback' => 'absint',
+						'validate_callback' => 'rest_validate_request_arg',
```
Comment (joehoyle) [inline inc/export/class-endpoint.php:54]: these are not needed when passing correct `type` -- I've seen this pattern some a few people though, do you know where it comes from?

File: inc/export/class-endpoint.php [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	 * @return true
+	 */
+	public function pre_serve_request( bool $served, $result, WP_REST_Request $request, WP_REST_Server $server ) : bool {
+		$params = $request->get_params();
+
+		if ( ! preg_match( '#/analytics/v1/export/(?P<date>\d{4}-\d{2}-\d{2})#', $request->get_route() ) || 'GET' !== $request->get_method() ) {
+			return $served;
+		}
+
+		if ( ! isset( $params['format'] ) || 'csv' !== $params['format'] ) {
```
Comment (joehoyle) [inline inc/export/class-endpoint.php:183]: does it make sense to read from `Accept` header too?

File: inc/export/class-endpoint.php [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	}
+
+	/**
+	 * Lock down the endpoint to only allow administrator users authenticated via application password to access the
+	 * date.
+	 *
+	 * @return bool True if the request has read access for the item, otherwise false.
+	 */
+	public function get_item_permissions_check() {
+		return current_user_can( 'manage_options' );
```
Comment (joehoyle) [inline inc/export/class-endpoint.php:321]: Assuming we don't have any meta caps in analytics?

File: inc/export/class-endpoint.php [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+		add_filter( 'rest_pre_serve_request', [ $this, 'pre_serve_request' ], 10, 4 );
+	}
+
+	/**
+	 * Registers the route.
+	 */
+	public function register_routes() : void {
+		register_rest_route(
+			'analytics/v1',
+			'export/(?P<date>\d{4}-\d{2}-\d{2})',
```
Comment (joehoyle) [inline inc/export/class-endpoint.php:41]: isn't an API always "exporting"? what's the data type? `events`?

**Review by joehoyle — COMMENTED** [review]
Looking good.

I don't think it's in here, but it might be valuable to map some ID fields to slug. E>g. audience IDs or Post IDs are pretty useless when exporting data. Having the audience name in the CSV along with it's ID would be valuable.

File: inc/export/class-endpoint.php (hunk 1 above)
Comment (roborourke) [reply inc/export/class-endpoint.php:54]: This was from the original code, not sure on the pattern itself. Maybe @fklein-lu would have more information

File: inc/export/class-endpoint.php (hunk 3 above)
Comment (roborourke) [reply inc/export/class-endpoint.php:321]: Not for analytics directly but we could add an alternative capability here or at least a filter in the meantime.

File: inc/export/class-endpoint.php (hunk 4 above)
Comment (roborourke) [reply inc/export/class-endpoint.php:41]: GET requests are I suppose. Yeah `events` is a fair description of the type 👍

File: inc/export/class-endpoint.php (hunk 2 above)
Comment (roborourke) [reply inc/export/class-endpoint.php:183]: I'm not sure, I thought `Accept` headers could contain multiple mime types. Can do though, shouldn't be hard to add.

Comment (rmccue) [reply inc/export/class-endpoint.php:183]: `Accept` can contain multiple MIME types, and it's a bit of a pain to parse, but it's likely worth doing. [RFC 7231 §5.3.2](https://tools.ietf.org/html/rfc7231#section-5.3.2) is the relevant part.

To summarise from my memory: `Accept` is a comma-separated list of multiple values (as all† HTTP headers are). Each value is specified as `<type>/<subtype>` which can optionally have a wildcard (`*`) as the subtype or type and subtype, and can be followed with a semi-colon and then an arbitrary number of `key=value` parameter pairs separated by semicolons. The main one of those parameters is `q` which indicates the weight (i.e. priority) with 1 being the highest.

i.e. `Accept: text/html, application/xhtml+xml, application/xml;q=0.9, image/webp, */*;q=0.8` is
* Give me `text/html`, `application/xhtml+xml`, or `image/webp` with highest weight (q=1)
* Give me `application/xml` as the fallback, with weight 0.9
* Failing those, give me anything with `q=0.8`

**Comment by roborourke** [comment]
@joehoyle re.

> I don't think it's in here, but it might be valuable to map some ID fields to slug. E>g. audience IDs or Post IDs are pretty useless when exporting data. Having the audience name in the CSV along with it's ID would be valuable.

I'm not sure if this is the best place to do that. I consciously opted not to use data points that are quite as mutable as post slugs by default but to be fair I was mainly thinking about usage within WP. I would be tempted to recommend as far as Data Studio usage is concerned creating additional data sources, eg. get a CSV of audiences, posts etc if you want to enrich the data. You do at least get the URL in the attributes.

Otherwise I'd aim to handle extending the default data added separately to this PR. In the interim you can filter the default data passed from the server side as well if you want more info in there.

File: inc/export/class-endpoint.php (hunk 4 above)
Comment (roborourke) [reply inc/export/class-endpoint.php]: Changed `export` to `events`

File: inc/export/class-endpoint.php (hunk 3 above)
Comment (roborourke) [reply inc/export/class-endpoint.php]: Added a filter for now but we can come back to this. I feel like we'd need a solid solution for role & cap management first.

File: inc/export/class-endpoint.php (hunk 2 above)
Comment (roborourke) [reply inc/export/class-endpoint.php]: OK, added. Seems to work ok. Cheers @rmccue !

**Comment by joehoyle** [comment]
Ok - in terms of what I'm looking for, for the integrate into data studio etc, I'd expect the following:

- Audience names so I don't have to map an internal ID to more complex combinations of CSVs etc
- Significantly higher event limits, I don't want to have to write a program to loop over pagination. (if you want prior work: I wrote a css/json streaming output for vast amounts of data on vantage-backend for the access logs endpoint). 

It might be that this PR doesn't need to do those things, 1 sounds like something that could come later. 2 does seem like an architectural decision though, that might be better to approach now. If you are streaming data with appropriate nginx buffer headers etc, you should be able to keep the connection open for more than 60 seconds and as long as you don't have memory leaks you shouldn't run out of memory.

Right now, this seems mostly like a REST API for events.. which is pretty expected and good n all... but I think there's a gap between that and the feature / use-case of being able to export data for use in other tools. The "include all the data you need, not just internal IDs" comes to that question.

From the code side, I think what you have makes a good rest api endpoint. I'll defer to @rmccue on a final review. My feedback is mostly around the use-case and usefulness.

**Comment by roborourke** [comment]
I was thinking streaming data would be a better approach, we’re a bit stuck in terms of how Elasticsearch and paging works though. I’ll see how difficult it is.

We can probably add the audience names to the endpoint data by default but that would be done in the JS. I guess my thought was that those audiences might be defined in the BI tool too.

File: inc/export/class-endpoint.php [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+			foreach ( $index_mappings as $mapping ) {
+				$fields = array_merge( $fields, array_keys( $mapping['mappings'] ) );
+				$fields = array_unique( $fields );
+			}
+
+			sort( $fields );
+
+			// Write columns headings line.
+			fputcsv( fopen( 'php://output', 'w' ), $fields ) . "\n";
```
Comment (rmccue) [inline inc/export/class-endpoint.php:169]: You're reopening output handles each time here; you should open the handle once. Also, after each call here, you should also call `fflush()`

Comment (roborourke) [reply inc/export/class-endpoint.php:169]: Okey doke, might need some updates where this pattern is used elsewhere 😬

**Comment by roborourke** [comment]
Ok, streaming appears to work locally at least. I haven't tested beyond 20k documents but it seems to be able to export them in less than 30 seconds anyway. Requesting the CSV format seems to automatically trigger a download in the browser anyway so that's good. We could just add a link to the appropriate URL in the backend somewhere in future.

**Comment by rmccue** [comment]
Can we test this early on production infrastructure? I want to ensure we're checking compatibility early.

**Comment by roborourke** [comment]
@rmccue we can merge this, make a new release of this plugin and add it to a new preview release of the Altis Analytics module

**Comment by roborourke** [comment]
Deployed to platform-test but the kinesis firehose doesn't seem to be delivering to the new ES domain yet.

**Comment by roborourke** [comment]
Had an issue with cloud not allowing accept headers. Is there a way to allow that for a particular endpoint or is that a CloudFront limitation we have between the load balancer and the web server?

**Review by joehoyle — COMMENTED** [review]
@roborourke wondering if / why you don't need to set the accel-buffering header to make Nginx not buffer the response (see https://github.com/humanmade/vantage-backend/blob/6a16494d60a7a6e0d622a376f3069604ae8788ff/content/mu-plugins/hm-stack-core/inc/namespace.php#L252)

Do we not need that, or were you testing the streaming some other way?

File: inc/utils/namespace.php [hunk 6]
```diff
@@ -109,16 +109,45 @@ function get_elasticsearch_url() : string {
...
 }
 
+/**
+ * Retrieve the elasticsearch version.
+ *
+ * Return the version string if found otherwise 'unknown'.
+ *
+ * @return string
+ */
+function get_elasticsearch_version() : string {
```
Comment (joehoyle) [inline inc/utils/namespace.php:119]: Could we have this set by cloud maybe in a constant too? It's a bit of a pain when we're pinging off requests to elastic search on potentially unrelated page loads.

**Comment by roborourke** [comment]
Ok! This works on cloud now with an application password, we can maybe look at adding a link somewhere in the admin to fetch it but not sure how best to authenticate from a link right now.

**Comment by roborourke** [comment]
@joehoyle the `X-Accel-Buffering` header is set in the `get_items()` method. I just took it from the example you sent me. Is there more I need to add to make it work?

File: inc/utils/namespace.php (hunk 6 above)
Comment (roborourke) [reply inc/utils/namespace.php:119]: Yeah if we update TAS to set it as an environment variable we could use that if it exists. Can you create an issue on the cloud module for that?

Comment (roborourke) [reply inc/utils/namespace.php]: Added an issue to the terraform repo instead as that's where we'd get the info from.

**Review by joehoyle — APPROVED** [review]
Fandabidozi

**Pull request merged**